### PR TITLE
Nvfuser guard patch

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes issue where CudaFusionGuard would return false on backward graph because `requires_grad` flag doesn't match.

This is due to the fact that autodiff uses GradMode switch to turn on/off requires_grad, which is not taken into consideration by nvfuser guard. We verified the implementation under `TensorType::matchTensor`.

- [x] Add python test to verify no fallback is observed

### CODE REVIEW

**Review by eellison — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/interface.cpp [hunk 1]
```diff
@@ -146,7 +149,8 @@ bool complyWith(
       (guard_tensor_type->device().has_value() &&
        (guard_tensor_type->device().value() != tensor.device())) ||
       (guard_tensor_type->requiresGrad().has_value() &&
-       guard_tensor_type->requiresGrad().value() != tensor.requires_grad())) {
+       guard_tensor_type->requiresGrad().value() !=
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/interface.cpp:152]: Thanks for making me feel better~

**Comment by eellison** [comment]
have some failing tests

**Comment by malfet** [comment]
Temporary reverting in order to resolve land race between this PR and https://github.com/pytorch/pytorch/commit/96c87992c65f5e6bb1bdd51791682dd837af99b4

**Comment by davidberard98** [comment]
sorry didn't notice this when I landed the ci stuff, @jjsjann123 could you rebase this and also change `RUN_CUDA` -> `RUN_NVFUSER` to avoid rocm failures?

**Comment by malfet** [comment]
Oh, so it needed to be reverted anyway ;)

**Comment by davidberard98** [comment]
yeah, either this or the CI changes would have needed to be reverted

**Comment by Chillee** [comment]
Will reland in https://github.com/pytorch/pytorch/pull/75303 if @jjsjann123 doesn't mind :)

**Comment by jjsjann123** [comment]
> Will reland in #75303 if @jjsjann123 doesn't mind :)

ohohoh, somehow haven't got back to this one. Thanks for taking care of it.

**Comment by Chillee** [comment]
np - useful for me :)

**Pull request closed without merging**